scanner/engine/master.cpp

```diff
@@ -370,40 +438,123 @@ grpc::Status MasterImpl::NextWork(grpc::ServerContext* context,
-      // More tasks left
-      task_sampler_.reset(new TaskSampler(
-          *table_metas_.get(), job_params_.task_set().tasks(next_task_)));
-      task_result_ = task_sampler_->validate();
-      if (task_result_.success()) {
-        samples_left_ = task_sampler_->total_samples();
-        next_task_++;
-        VLOG(1) << "Tasks left: " << num_tasks_ - next_task_;
+  if (!worker_active_.at(node_info->node_id())) {
+    // Worker is not active
+    new_work->mutable_io_item()->set_item_id(-1);
+    return grpc::Status::OK;
+  }
+
+  // If we do not have any outstanding work, try and create more
+  if (unallocated_task_samples_.empty()) {
+    // If we have no more samples for this task, try and get another task
+    if (next_sample_ == num_samples_) {
+      // Check if there are any tasks left
+      if (next_task_ < num_tasks_ && task_result_.success()) {
+        // More tasks left
+        auto sampler = new TaskSampler(
+            *table_metas_.get(), job_params_.task_set().tasks(next_task_));
+        task_result_ = sampler->validate();
+        if (task_result_.success()) {
+          next_sample_ = 0;
+          num_samples_ = sampler->total_samples();
+          next_task_++;
+          VLOG(1) << "Tasks left: " << num_tasks_ - next_task_;
+          task_samplers_[next_task_ - 1].reset(sampler);
+        } else {
+          delete sampler;
+        }
       }
-    } else {
-      // No more tasks left
-      new_work->mutable_io_item()->set_item_id(-1);
-      return grpc::Status::OK;
+    }
+
+    // Create more work if possible
+    if (next_sample_ < num_samples_) {
+      i64 current_task = next_task_ - 1;
+      i64 current_sample = next_sample_;
+
+      unallocated_task_samples_.push_front(
+          std::make_tuple(current_task, current_sample));
+      task_sampler_samples_left_[current_task]++;
+      next_sample_++;
     }
   }
-  if (!task_result_.success()) {
+
+  if (unallocated_task_samples_.empty()) {
+    // No more work
     new_work->mutable_io_item()->set_item_id(-1);
     return grpc::Status::OK;
   }
 
-  assert(samples_left_ > 0);
-  task_result_ = task_sampler_->next_work(*new_work);
+  // Grab the next task sample
+  std::tuple<i64, i64> task_sample_id = unallocated_task_samples_.back();
+  unallocated_task_samples_.pop_back();
+
+  // Get task sampler for our task sample
+  assert(next_sample_ <= num_samples_);
+  auto& sampler = task_samplers_.at(std::get<0>(task_sample_id));
+
+  // Get the task sample
+  task_result_ = sampler->sample_at(std::get<1>(task_sample_id), *new_work);
   if (!task_result_.success()) {
+    // Task sampler failed for some reason
     new_work->mutable_io_item()->set_item_id(-1);
     return grpc::Status::OK;
   }
+  new_work->mutable_load_work()->set_job_index(std::get<0>(task_sample_id));
+
+  // Track sample assigned to worker
+  active_task_samples_[node_info->node_id()].insert(task_sample_id);
+  worker_histories_[node_info->node_id()].tasks_assigned += 1;
```

Comment:
I currently don't use it but I want to use it to keep track of how much worker is assigned to each worker (or even calculate their throughput). I was imagining displaying this on a browser page or as an enhanced progress bar.